Classes/Event/EventWithMetadata.php

```diff
@@ -0,0 +1,56 @@
+<?php
+namespace Neos\EventSourcing\Event;
+
+/*
+ * This file is part of the Neos.EventSourcing package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+/**
+ * Event wrapper which provides metadata additional to the event
+ */
+final class EventWithMetadata implements EventInterface
+{
+    /**
+     * @var EventInterface
+     */
+    private $event;
+
+    /**
+     * @var array
+     */
+    private $metadata;
+
+    /**
+     * EventWithMetadata constructor.
+     *
+     * @param EventInterface $event
+     * @param array $metadata
+     */
+    public function __construct(EventInterface $event, array $metadata)
```

Comment:
As mentioned above, I'd suggest to have an `EventWithMetadataInterface`(extending `EventInterface`).
If you agree, this constructor should probably check if the `$event` implements that interface and merge the `$metadata` in that case.
This would allow to augment events while publishing them but they could also specify their own metadata

Reply:
The practical issue I see in that is that the even will also contain the metadata and that would be serialized as a payload. Do we really want that?

Reply:
I would not put it in the payload but read it from `RawEvent::getMetadata()`. But you're right, that possibly raises some more issues.